lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FieldToPointFunctionalChaosAlgorithm.cxx

```diff
@@ -0,0 +1,438 @@
+    const Bool centered = true;
+    KarhunenLoeveSVDAlgorithm algo(inputProcessSampleI, threshold_, centered);
+    algo.setNbModes(nbModes_);
+    algo.run();
+    eigenValues.add(algo.getResult().getEigenvalues());
+    klResultCollection[i] = algo.getResult();
+    LOGINFO(OSS() << "block #" << i << " indices=" << blockIndices_[i] << " ev size=" << algo.getResult().getEigenvalues().getDimension());
+  }
+
+  if (recompress_)
+  {
+    const Scalar cumulatedVariance = eigenValues.norm1();
+    std::sort(eigenValues.begin(), eigenValues.end(), [](const Scalar a, const Scalar b) { return a > b; });
+    const UnsignedInteger nbModesMax = std::min(nbModes_, eigenValues.getSize());
+    // Find the cut-off in the eigenvalues
+    UnsignedInteger K = 0;
+    Scalar selectedVariance = 0.0; // sum of eigenvalues selected after cut-off is applied
+    do
+    {
+      selectedVariance += eigenValues[K];
+      ++ K;
+    } while ((K < nbModesMax) && (selectedVariance < (1.0 - threshold_) * cumulatedVariance));
+    LOGINFO(OSS() << "Selected " << K << " eigenvalues out of " << eigenValues.getSize() << " computed");
+
+    const Scalar lambdaCut = eigenValues[K];
+    for (UnsignedInteger i = 0; i < blockIndices_.getSize(); ++ i)
+    {
+      UnsignedInteger Ki = 1;// retain at least 1 mode
+      Point eigenValuesI(klResultCollection[i].getEigenvalues());
+      Scalar selectedVarianceI = eigenValuesI[0];
+      const Scalar cumulatedVarianceI = eigenValuesI.norm1();
+      while ((Ki < eigenValuesI.getSize()) && (eigenValuesI[Ki] > lambdaCut))
+      {
+        selectedVarianceI += eigenValuesI[Ki];
+        ++ Ki;
+      }
+      eigenValuesI.resize(Ki);
+      Collection<Function> modesI(klResultCollection[i].getModes());
+      modesI.resize(Ki);
+      CovarianceModel covarianceI(RankMCovarianceModel(eigenValuesI, modesI));
+      ProcessSample modesAsProcessSampleI(klResultCollection[i].getModesAsProcessSample());
+      modesAsProcessSampleI.erase(Ki, modesAsProcessSampleI.getSize());
+      MatrixImplementation projectionMatrixI(*klResultCollection[i].getProjectionMatrix().getImplementation());
+      projectionMatrixI.resize(Ki, projectionMatrixI.getNbColumns());
+      
+      klResultCollection[i] = KarhunenLoeveResult(covarianceI, klResultCollection[i].getThreshold(), eigenValuesI, modesI, modesAsProcessSampleI, projectionMatrixI, selectedVarianceI/cumulatedVarianceI);
+    }
+  }
+
+  // the global input projection stacks projections of each block of variables  
+  Sample modesSample(size, 0);
+  Collection<FieldToPointFunction> projectionCollection(blockIndices_.getSize());
+  for (UnsignedInteger i = 0; i < blockIndices_.getSize(); ++ i)
+  {
+    const KarhunenLoeveProjection projectionI(klResultCollection[i]);
+    projectionCollection[i] = projectionI;
+    const ProcessSample inputProcessSampleI(inputProcessSample.getMarginal(blockIndices_[i]));
+    modesSample.stack(projectionI(inputProcessSampleI));
```

Comment:
Indeed. As the independence structure is given by the user through the blocks of independent marginal processes, we know that the corresponding KL coefficients are independent.
The code should produce a BlockIndependentDistribution, with marginals estimated the way they are and for each block,  the copula should be estimated using the following rule:
1) an IndependentCopula if the independence test is successful,
2) a NormalCopula if a specific key is set in ResourceMap,
3) an EmpiricalBernsteinCopula otherwise. The use of a BetaCopula (i.e. an EmpiricalBernsteinCopula with a bin number equal to the sample size) is a good practice for input samples resulting from the transport of a low dimensional random vector. For a truly infinite dimensional process, the use of BernsteinCopulaFactory is mandatory but could be added in a future revision of the code.

For a diffuse copula (i.e. a copula with a PDF) the AMISE value decreases exponentially slowly with the dimension, hence the interest in using low dimensional blocks. But for the BetaCopula parameterization the only contributor to the quadratic error is the variance (at least to the first order) so there is no dependence to the dimension

Regarding the performance of the isoprobabilistic transformation, the following benchmark suggests that the performance should be improved for the inverse transformation but not for the direct transformation, which is strange to me... BTW the evaluation over a sample is sequential and could be parallelized.

```
from openturns import *
from time import time

size = 10000
big = 100
data = Normal(8).getSample(size)
distribution_full = ComposedDistribution([Normal()]*8, EmpiricalBernsteinCopula(data, size))
iso_full = distribution_full.getIsoProbabilisticTransformation()
inv_iso_full = distribution_full.getInverseIsoProbabilisticTransformation()
test = Normal(8).getSample(big)
t0 = time()
res = iso_full(test)
t1 = time()
print("iso_full=", big / (t1 - t0), "evals/s")
t0 = time()
res = inv_iso_full(test)
t1 = time()
print("inv_iso_full=", big / (t1 - t0), "evals/s")
blocks = list()
for i in range(4):
    blocks.append(ComposedDistribution([Normal()]*2, EmpiricalBernsteinCopula(data.getMarginal([2*i, 2*i+1]), size)))
distribution_block = BlockIndependentDistribution(blocks)
iso_block = distribution_block.getIsoProbabilisticTransformation()
inv_iso_block = distribution_block.getInverseIsoProbabilisticTransformation()
test = Normal(8).getSample(big)
t0 = time()
res = iso_block(test)
t1 = time()
print("iso_block=", big / (t1 - t0), "evals/s")
t0 = time()
res = inv_iso_block(test)
t1 = time()
print("inv_iso_block=", big / (t1 - t0), "evals/s")
```
gives:
```
iso_full= 127.4240485889736 evals/s
inv_iso_full= 5.6209173993688495 evals/s
iso_block= 98.58922324421536 evals/s
inv_iso_block= 17.734506200834602 evals/s
```

Reply:
Thanks. This probably means within `BlockIndependentDistribution`, must of time is due to creation/destruction of `Point`

Reply:
ok to add a resourcemap key for Normal/Beta copulas